### PR TITLE
vm/agent_groups

### PR DESCRIPTION
**Pull request by milanaghera-tenable**
# Description

Provided support for v3 endpoint for agent_groups API.
- convert the type of agent and group id to UUID.
- update the examples in sphinx docs for all the methods. 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Python Version(s) Tested:
* Tenable.sc version (if necessary):

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tenable/io/v3/vm/api.py [hunk 1]
```diff
@@ -37,6 +39,14 @@ class VulnerabilityManagement(APIEndpoint):  # noqa: PLR0904
     i.e assets, agents, scanners etc.
     '''
 
+    @property
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/api.py:42]: Please follow the ASCII order. So, Add agent_groups after agent_config, similar to the order in imports.

Comment (milanaghera-tenable) [reply tenable/io/v3/vm/api.py]: update the changes as per comment.

File: tenable/io/v3/vm/agent_groups.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    def search(self):
+        NotImplemented('Search and Filter functionality '
+                       'will be updated later.')
+
+    def task_status(self, group_id: int, task_uuid: str) -> Dict:
+        '''
+        Retrieves the current status of a bulk task.
+
+        :devportal:`bulk-operations: bulk-agent-group-status
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups.py:162]: @milanaghera-tenable Could you please verify once that if you divide this line into two, is it properly being reflected on Sphinx?

File: tenable/io/v3/vm/agent_groups.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+        <bulk-task-agent-group-status>`
+
+        Args:
+            group_id (int): The id of the group
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups.py:166]: group_id will now be of type UUID (str) as per latest v3 spec.

File: tenable/io/v3/vm/agent_groups.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+    def delete_agent(self, group_id: int, *agent_ids: int)\
+            -> Union[None, Dict]:
+        '''
+        Delete one or many agents from an agent group.
+
+        :devportal:`agent-groups: delete-agent <agent-groups-delete-agent>`
+
+        Args:
+            group_id (int): The id of the agent group to remove the agent from
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups.py:121]: Type of group_id and agent_ids should now be updated to UUID (str).

File: tenable/io/v3/vm/agent_groups.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        return self._post(f'', json={'name': name})
+
+    def delete(self, group_id: int) -> None:
+        '''
+        Delete an agent group.
+
+        :devportal:`agent-groups: delete <agent-groups-delete>`
+
+        Args:
+            group_id (int): The id of the agent group to delete
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups.py:103]: group_id should be UUID (str).

File: tenable/io/v3/vm/agent_groups.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+            name (str): The name of the agent group
+
+        Returns:
+            :obj:`dict`:
+                The dictionary object representing the newly minted agent group
+
+        Examples:
+            >>> group = tio.v3.vm.agent_groups.create('New Agent Group')
+        '''
+        return self._post(f'', json={'name': name})
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups.py:94]: You can simply write:
return self._post(json={'name': name})

File: tenable/io/v3/vm/agent_groups.py [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+            )
+
+    def configure(self, group_id: int, name: str) -> Dict:
+        '''
+        Renames an existing agent group.
+
+        :devportal:`agent-groups: configure <agent-groups-configure>`
+
+        Args:
+            group_id (int): The id of the group
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups.py:67]: group_id => UUID (str)

File: tenable/io/v3/vm/agent_groups.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    _conv_json: bool = True
+
+    def add_agent(self, group_id: int, *agent_ids: int) -> Union[None, Dict]:
+        '''
+        Adds an agent or multiple agents to the agent group specified.
+
+        :devportal:`agent-groups: add-agent <agent-groups-add-agent>`
+
+        Args:
+            group_id (int): The id of the group
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups.py:30]: group_id and agent_id => UUID (str)

File: tests/io/v3/vm/test_agent_groups.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+import responses
+
+BASE_URL = 'https://cloud.tenable.com/api/v3/agent-groups'
+
+
+@responses.activate
+def test_add_agent_with_single_agent_id(api):
+    '''Test case for add_agent method with sigle agent id'''
+    group_id: int = 470009
+    agent_id: int = 5
```
Comment (dhvanilshah-tenable) [inline tests/io/v3/vm/test_agent_groups.py:16]: type should be UUID everywhere for group_id and agent_id

File: tests/io/v3/vm/test_agent_groups.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+
+@responses.activate
+def test_add_agent_with_multiple_agent_id(api):
+    '''Test case for add_agent method with multiple agent id'''
+    group_id: int = 470009
+    payload = {'items': [i for i in (1, 2, 3, 4)]}
+    test_response = {
+        'task_id': '07a665f4-6e09-444b-b9ce-ed5ccf5c193b',
+        'container_uuid': 'cfdabb09-6aef-481d-b28f-aecb1c38f297',
```
Comment (dhvanilshah-tenable) [inline tests/io/v3/vm/test_agent_groups.py:32]: This should be container_id vs container_uuid currently.

File: tests/io/v3/vm/test_agent_groups.py [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+
+@responses.activate
+def test_task_status(api):
+    '''Test case for task_status method'''
+    group_id: int = 470009
+    task_id: str = '02683e5e-4b7c-4d2b-99a1-cde1ea0940d9'
+    test_response = {
+        'task_id': '02683e5e-4b7c-4d2b-99a1-cde1ea0940d9',
+        'container_uuid': 'cfdabb09-6aef-481d-b28f-aecb1c38f297',
```
Comment (dhvanilshah-tenable) [inline tests/io/v3/vm/test_agent_groups.py:161]: container_id vs container_uuid

File: tenable/io/v3/vm/agent_groups.py (hunk 2 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups.py:162]: yes, it is. I have checked that.

File: tenable/io/v3/vm/agent_groups.py (hunk 8 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups.py:30]: updated.

File: tenable/io/v3/vm/agent_groups.py (hunk 7 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups.py:67]: updated.

File: tenable/io/v3/vm/agent_groups.py (hunk 6 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups.py:94]: done.

File: tenable/io/v3/vm/agent_groups.py (hunk 5 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups.py:103]: done.

File: tenable/io/v3/vm/agent_groups.py (hunk 4 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups.py:121]: done.

File: tenable/io/v3/vm/agent_groups.py (hunk 3 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups.py:166]: done,

File: tests/io/v3/vm/test_agent_groups.py (hunk 9 above)
Comment (milanaghera-tenable) [reply tests/io/v3/vm/test_agent_groups.py]: Done.

File: tests/io/v3/vm/test_agent_groups.py (hunk 10 above)
Comment (milanaghera-tenable) [reply tests/io/v3/vm/test_agent_groups.py]: Done.

File: tests/io/v3/vm/test_agent_groups.py (hunk 11 above)
Comment (milanaghera-tenable) [reply tests/io/v3/vm/test_agent_groups.py]: Done.

File: tenable/io/v3/vm/agent_groups/api.py [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+'''
+from typing import Dict, Union
+from uuid import UUID
+
+from tenable.io.v3.base.endpoints.explore import ExploreBaseEndpoint
+from tenable.io.v3.vm.agent_groups.schema import AgentGroupsBaseSchema
+
+
+class AgentGroupsAPI(ExploreBaseEndpoint):
+    _path: str = f'api/v3/agent-groups'
```
Comment (harshshah-tenable) [inline tenable/io/v3/vm/agent_groups/api.py:22]: No need for a formatted string

File: tenable/io/v3/vm/agent_groups/api.py [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+        Returns:
+            :obj:`dict`
+
+        Examples:
+            >>> tio.v3.vm.agent_groups.configure(
+            ...    'ef62870e-fe2f-4ba9-98b7-43d3a53ffe85',
+            ...    'New Name'
+            ... )
+        '''
+        return self._put(f'{group_id}', json={'name': name})
```
Comment (harshshah-tenable) [inline tenable/io/v3/vm/agent_groups/api.py:91]: Can we validate this payload with schema as well similar to create action

**Review by harshshah-tenable — APPROVED** [review]

File: tenable/io/v3/vm/agent_groups/api.py (hunk 12 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups/api.py]: updated

File: tenable/io/v3/vm/agent_groups/api.py (hunk 13 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups/api.py]: yes, we can. updated as per comment.

File: tenable/io/v3/vm/agent_groups/api.py [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+            ... )
+        '''
+        if len(agent_ids) <= 1:
+            # if there is only 1 agent id, we will perform a singular add.
+            self._put(f'{group_id}/agents/{agent_ids[0]}')
+        else:
+            # If there are many agent_ids, then we will want to perform a
+            # bulk operation.
+            payload: dict = {'items': [i for i in agent_ids]}
+            schema = AgentGroupsBaseSchema(only=['items'])
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups/api.py:65]: Is there a need to use only keyword?

File: tenable/io/v3/vm/agent_groups/api.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+            :obj:`dict`
+
+        Examples:
+            >>> tio.v3.vm.agent_groups.configure(
+            ...    'ef62870e-fe2f-4ba9-98b7-43d3a53ffe85',
+            ...    'New Name'
+            ... )
+        '''
+        payload: dict = {'name': name}
+        schema = AgentGroupsBaseSchema(only=['name'])
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups/api.py:92]: Is there a need to use only keyword?

File: tenable/io/v3/vm/agent_groups/api.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+
+        Returns:
+            :obj:`dict`:
+                The dictionary object representing the newly minted agent group
+
+        Examples:
+            >>> group = tio.v3.vm.agent_groups.create('New Agent Group')
+        '''
+        payload: dict = {'name': name}
+        schema = AgentGroupsBaseSchema(only=['name'])
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups/api.py:113]: Is there a need to use only keyword?

File: tenable/io/v3/vm/agent_groups/api.py [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+            :obj:`None`
+
+        Examples:
+            >>> tio.v3.vm.agent_groups.delete(
+            ...     'ef62870e-fe2f-4ba9-98b7-43d3a53ffe85'
+            ... )
+        '''
+        self._delete(f'{group_id}')
+
+    def delete_agent(self, group_id: UUID, *agent_ids: UUID) \
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups/api.py:136]:
```
    def delete_agent(self,
                     group_id: UUID, 
                     *agent_ids: UUID) -> Union[None, Dict]:
```

File: tenable/io/v3/vm/agent_groups/api.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+            ... )
+        '''
+        if len(agent_ids) <= 1:
+            # if only a singular agent_id was passed, then we will want to
+            self._delete(f'{group_id}/agents/{agent_ids[0]}')
+        else:
+            # if multiple agent ids were requested to be deleted, then we will
+            # call the bulk deletion API.
+            payload: dict = {'items': [i for i in agent_ids]}
+            schema = AgentGroupsBaseSchema(only=['items'])
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups/api.py:178]: Do we require only keyword?

File: tenable/io/v3/vm/agent_groups/api.py [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+        Args:
+            group_id (UUID): The unique identifier of the group
+            task_id (UUID): The unique identifier of the task
+
+        Returns:
+            :obj:`dict`:
+                Task resource
+
+        Examples:
+            >>> item = tio.v3.vm.agent_groups.add_agnet(
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/agent_groups/api.py:209]: typo => tio.v3.vm.agent_groups.add_agent(

File: tests/io/v3/vm/agent_groups/test_agent_groups_api.py [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+        match=[responses.matchers.json_params_matcher(payload)],
+        json=test_repsonse
+    )
+    res = api.v3.vm.agent_groups.delete_agent(
+        group_id,
+        '57b74c0a-5d95-11ec-bf63-0242ac130002',
+        '57b74e58-5d95-11ec-bf63-0242ac130002',
+        '57b74f66-5d95-11ec-bf63-0242ac130002'
+    )
+    assert isinstance(res, dict)
```
Comment (dhvanilshah-tenable) [inline tests/io/v3/vm/agent_groups/test_agent_groups_api.py:199]: We should try to compare a couple of keys as well.

File: tests/io/v3/vm/agent_groups/test_agent_groups_api.py [hunk 21]
```diff
@@ -0,0 +1,242 @@
...
+        'end_time': 1638807034730,
+        'total_work_units_completed': 0
+    }
+    responses.add(
+        responses.GET,
+        re.compile(f'{BASE_URL}/{group_id}/agents/_bulk/{task_id}'),
+        json=test_response
+    )
+    res = api.v3.vm.agent_groups.task_status(group_id, task_id)
+    assert isinstance(res, dict)
```
Comment (dhvanilshah-tenable) [inline tests/io/v3/vm/agent_groups/test_agent_groups_api.py:224]: Assert couple of keys.

File: tests/io/v3/vm/agent_groups/test_agent_groups_schema.py [hunk 22]
```diff
@@ -0,0 +1,41 @@
...
+    Test the agent_groups schema with name
+    '''
+    name: str = 'sample name'
+    payload = {
+        'name': name
+    }
+    test_resp = {
+        'name': name
+    }
+    schema = AgentGroupsBaseSchema(only=['name'])
```
Comment (dhvanilshah-tenable) [inline tests/io/v3/vm/agent_groups/test_agent_groups_schema.py:18]: According to me, we don't need to use only keyword in the complete Agent Group API.

File: tenable/io/v3/vm/agent_groups/api.py (hunk 14 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups/api.py:65]: Resolved.

File: tenable/io/v3/vm/agent_groups/api.py (hunk 15 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups/api.py:92]: Resolved.

File: tenable/io/v3/vm/agent_groups/api.py (hunk 16 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups/api.py:113]: Resolved

File: tenable/io/v3/vm/agent_groups/api.py (hunk 17 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups/api.py:136]: Updated.

File: tenable/io/v3/vm/agent_groups/api.py (hunk 18 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups/api.py:178]: Resolved.

File: tenable/io/v3/vm/agent_groups/api.py (hunk 19 above)
Comment (milanaghera-tenable) [reply tenable/io/v3/vm/agent_groups/api.py:209]: Updated.

File: tests/io/v3/vm/agent_groups/test_agent_groups_api.py (hunk 20 above)
Comment (milanaghera-tenable) [reply tests/io/v3/vm/agent_groups/test_agent_groups_api.py:199]: Added.

File: tests/io/v3/vm/agent_groups/test_agent_groups_api.py (hunk 21 above)
Comment (milanaghera-tenable) [reply tests/io/v3/vm/agent_groups/test_agent_groups_api.py:224]: Added.

File: tests/io/v3/vm/agent_groups/test_agent_groups_schema.py (hunk 22 above)
Comment (milanaghera-tenable) [reply tests/io/v3/vm/agent_groups/test_agent_groups_schema.py:18]: Updated.

File: tests/io/v3/vm/agent_groups/test_agent_groups_api.py [hunk 23]
```diff
@@ -0,0 +1,248 @@
...
+    payload = {
+        'items': [
+            i for i in (
+                '57b74c0a-5d95-11ec-bf63-0242ac130002',
+                '57b74e58-5d95-11ec-bf63-0242ac130002',
+                '57b74f66-5d95-11ec-bf63-0242ac130002'
+            )
+        ]
+    }
+    test_repsonse = {
```
Comment (dhvanilshah-tenable) [inline tests/io/v3/vm/agent_groups/test_agent_groups_api.py:183]: typo test_repsonse  => test_response

File: tests/io/v3/vm/agent_groups/test_agent_groups_api.py (hunk 20 above)
Comment (dhvanilshah-tenable) [reply tests/io/v3/vm/agent_groups/test_agent_groups_api.py:201]: @milanaghera-tenable it should be 
assert test_response['container_id'] == res['container_id']

**Review by dhvanilshah-tenable — APPROVED** [review]

**Review by tomeara — APPROVED** [review]

**Pull request merged**